File: variable_selection/c1_Feature_ppe_mrmr_seq.m
function c1_Feature_ppe_mrmr_seq (X0,GQ)
addpath ("D:\Dropbox\Naresh_Sahu\Post Monsoon ANN\Codes_today")
load PPE_mean.mat meanloss

[idxmrmr,~] = fsrmrmr(X0,GQ);
[~, id_ann]= sort(meanloss{2},'descend');
[~, id_RF]= sort(meanloss{3},'descend');


mysequential(X0,GQ);

feature_selection_ppe_mrmr(idxmrmr,X0,GQ,'mrmr','features_mrmr.mat')
feature_selection_ppe_mrmr([id_ann id_RF],X0,GQ,'ppe','features_ppe.mat')

disp('')


function mysequential(Xtr,Ytr)
output=c2_Bayesopt_rf_for_varselect(Xtr,Ytr);
R2cvf=output{2};
n=size(Xtr,2);selected=1:n;
% R2cvf=0.9215;
% selected=[1 2 4 11 12];

R2cvi=0;
filename='Sequential.mat';

rf_history={}; %#ok<NASGU>
while R2cvf>0.90*R2cvi
    count=numel(selected);
    result=cell(count,3);tempMod=cell(count,1);
    for i=1:count
        v=selected;
        v(i)=[];
        %rng(1);

        %output=c2_Bayesopt_ann_for_varselect(Xtr(:,v),Ytr);
        [output]=c2_Bayesopt_rf_for_varselect(Xtr(:,v),Ytr);

        %mdl_R2=output{1};mdl_CV_R2=output{2};mdl=output{3};
        result(i,:)={v,output{1:2}};tempMod(i,:)=output(3);
    end
    [~, il]=sort([result{:,3}],'descend');%% sorted cell array on CVt
    result=result(il,:); %% sorted
    tempMod=tempMod(il,:);
    Newv=result{1,1};newR2=result{1,2};newCVR2=result{1,3};newMdl=tempMod(1,1);
    if newCVR2>0.90*R2cvf
        R2cvi=R2cvf;
        R2cvf=newCVR2;
        selected=Newv;
        newHistory={selected,newR2,newCVR2,newMdl};
        if isfile(filename)
            if ismember('rf_history', who('-file', filename))
                load (filename,'rf_history')
                rf_history=[rf_history;newHistory];
            else
                rf_history=newHistory;
            end
            save (filename,'rf_history','-append')
        else
            rf_history=newHistory;
            save (filename,'rf_history')
        end
    else
        break
    end
end


function feature_selection_ppe_mrmr(ppe,X0,GQ,Method,filename)

if size(ppe,2)==2
    id_ann=ppe(:,1);
    id_RF=ppe(:,2);
else
    id_ann=ppe;
    id_RF=ppe;
end



if exist(filename,'file')==2
    load (filename,'myT_RF','myT_ann')
else
    varNames={'Method','Model','Size','variables','R2','R2CV'};
    myT_RF = table('Size',[0 6],'VariableNames',varNames,'VariableTypes',{'cell','cell','double','cell','double','double'});
    myT_ann = table('Size',[0 6],'VariableNames',varNames,'VariableTypes',{'cell','cell','double','cell','double','double'});
end

for i=1:10

    vars_ann=id_ann(1:2+i); %% starts from top 3 variabes and increasing
    [output_ann]=c2_Bayesopt_ann_for_varselect(X0{:,vars_ann},GQ);
    R2_ann=output_ann{1};R2CV_ann=output_ann{2};Mdl_ann=output_ann{3};
    newRow_ann = {Method,{Mdl_ann}, numel(vars_ann), {vars_ann}, R2_ann, R2CV_ann};
    myT_ann=[myT_ann;newRow_ann]; %#ok<AGROW>

    vars_RF=id_RF(1:2+i); %% starts from top 3 variabes and increasing
    [output_rf]=c2_Bayesopt_rf_for_varselect(X0{:,vars_RF},GQ);
    R2_rf=output_rf{1};R2CV_rf=output_rf{2};Mdl_rf=output_rf{3};
    newRow_rf = {Method,{Mdl_rf},numel(vars_RF), {vars_RF}, R2_rf, R2CV_rf};
    myT_RF=[myT_RF;newRow_rf]; %#ok<AGROW>

    save (filename,'myT_RF', 'myT_ann')
end

File: python_code/variable_selection/c1_Feature_ppe_mrmr_seq.py
import numpy as np
from sklearn.feature_selection import mutual_info_regression
import os
import sys


def mysequential(Xtr, Ytr):
    from variable_selection.c2_Bayesopt_rf_for_varselect import c2_Bayesopt_rf_for_varselect

    output = c2_Bayesopt_rf_for_varselect(Xtr, Ytr)
    R2cvf = output[1]
    n = Xtr.shape[1]
    selected = list(range(n))

    R2cvi = 0
    filename = 'Sequential.npz'
    base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    save_dir = os.path.join(base_dir, 'saved_models')
    os.makedirs(save_dir, exist_ok=True)
    filepath = os.path.join(save_dir, filename)

    rf_history = []

    while R2cvf > 0.90 * R2cvi:
        count = len(selected)
        result = []

        for i in range(count):
            v = selected.copy()
            v.pop(i)
            output = c2_Bayesopt_rf_for_varselect(Xtr[:, v], Ytr)
            result.append({'v': v, 'R2': output[0], 'CVR2': output[1]})

        result.sort(key=lambda x: x['CVR2'], reverse=True)
        Newv = result[0]['v']
        newR2 = result[0]['R2']
        newCVR2 = result[0]['CVR2']

        if newCVR2 > 0.90 * R2cvf:
            R2cvi = R2cvf
            R2cvf = newCVR2
            selected = Newv
            newHistory = {'selected': selected, 'R2': newR2, 'CVR2': newCVR2}
            rf_history.append(newHistory)
        else:
            break

    rf_history_arr = np.array(rf_history)
    np.savez(filepath, rf_history=rf_history_arr)
    return rf_history


def feature_selection_ppe_mrmr(ppe, X0, GQ, Method, filename):
    base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    save_dir = os.path.join(base_dir, 'saved_models')
    os.makedirs(save_dir, exist_ok=True)
    filepath = os.path.join(save_dir, filename)

    from variable_selection.c2_Bayesopt_ann_for_varselect import c2_Bayesopt_ann_for_varselect
    from variable_selection.c2_Bayesopt_rf_for_varselect import c2_Bayesopt_rf_for_varselect

    if ppe.ndim == 2 and ppe.shape[1] == 2:
        id_ann = ppe[:, 0]
        id_RF = ppe[:, 1]
    else:
        id_ann = ppe.flatten()
        id_RF = ppe.flatten()

    try:
        data = np.load(filepath, allow_pickle=True)
        myT_RF = data['myT_RF'].tolist()
        myT_ann = data['myT_ann'].tolist()
    except (FileNotFoundError, IOError):
        myT_RF = []
        myT_ann = []

    X0_arr = X0 if isinstance(X0, np.ndarray) else X0.values

    for i in range(10):
        vars_ann = id_ann[:2 + i]
        vars_ann = vars_ann[vars_ann < X0_arr.shape[1]]
        if len(vars_ann) == 0:
            continue
        output_ann = c2_Bayesopt_ann_for_varselect(X0_arr[:, vars_ann], GQ)
        newRow_ann = {'Method': Method, 'Model': output_ann[2], 'Size': len(vars_ann),
                       'variables': vars_ann, 'R2': output_ann[0], 'R2CV': output_ann[1]}
        myT_ann.append(newRow_ann)

        vars_RF = id_RF[:2 + i]
        vars_RF = vars_RF[vars_RF < X0_arr.shape[1]]
        if len(vars_RF) == 0:
            continue
        output_rf = c2_Bayesopt_rf_for_varselect(X0_arr[:, vars_RF], GQ)
        newRow_rf = {'Method': Method, 'Model': output_rf[2], 'Size': len(vars_RF),
                      'variables': vars_RF, 'R2': output_rf[0], 'R2CV': output_rf[1]}
        myT_RF.append(newRow_rf)

    np.savez(filepath, myT_RF=np.array(myT_RF, dtype=object),
             myT_ann=np.array(myT_ann, dtype=object))


def c1_Feature_ppe_mrmr_seq(X0, GQ):
    base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    save_dir = os.path.join(base_dir, 'saved_models')

    X0_arr = X0 if isinstance(X0, np.ndarray) else X0.values

    ppe_data = np.load(os.path.join(save_dir, 'PPE_mean.npz'), allow_pickle=True)
    meanloss = ppe_data['meanloss']

    idxmrmr = np.argsort(mutual_info_regression(X0_arr, GQ))[::-1]

    id_ann = np.argsort(meanloss[1])[::-1]
    id_RF = np.argsort(meanloss[2])[::-1]

    mysequential(X0_arr, GQ)

    feature_selection_ppe_mrmr(idxmrmr, X0_arr, GQ, 'mrmr', 'features_mrmr.npz')

    ppe_indices = np.column_stack([id_ann, id_RF])
    feature_selection_ppe_mrmr(ppe_indices, X0_arr, GQ, 'ppe', 'features_ppe.npz')